docs/content/preview/api/ysql/user-defined-subprograms-and-anon-blocks/language-plpgsql-subprograms/plpgsql-syntax-and-semantics/executable-section/loop-exit-continue-statements/query-for-loop.md

```diff
@@ -0,0 +1,370 @@
+
+In most cases, you'll simply use ordinary variables or formal arguments (_i.e._ not a _record_) for the _loop variable(s)_ like this:
+
+```plpgsql
+create function s.g(v_lo in int, v_hi in int)
+  returns table(k_out int, v_out int)
+  set search_path = pg_catalog, pg_temp
+  language plpgsql
+as $body$
+begin
+  for k_out, v_out in (
+    with
+      c as (update s.t set v = v + 3 where k between v_lo and v_hi returning k, v)
+    select k, v from c
+    order by k
+    )
+  loop
+    return next;
+  end loop;
+end;
+$body$;
+
+select k_out, v_out from s.g(3, 7);
+```
+
+This is the result:
+
+```output
+ k_out | v_out 
+-------+-------
+     3 |    13
+     4 |    18
+     5 |    23
+```
+
+This example makes another point. The _query for loop_ cannot use an _insert... returning_, an _update... returning_, or a _delete... returning_ statement directly. (The attempt causes a syntax error.) However, the rules of top-level SQL _do_ allow a _select_ statement to use a _common table expression_ (_a.k.a. CTE_) that is defined with such a DML statement that uses a _returning_ clause. See the section [WITH clause—SQL syntax and semantics](../../../../../../the-sql-language/with-clause/with-clause-syntax-semantics/). You can therefore take advantage of this to use a _query for loop_ to iterate over the result set that the _returning_ clause of a DML statement defines.
+
+## "Query for loop" with a bound refcursor variable
+
+See the subsection [Semantics for the "plpgsql_bound_refcursor_declaration" rule](../../../declaration-section/#plpgsql-bound-refcursor-declaration) on the [declaration section](../../../declaration-section/) page.
+
+Try this:
+
+```plpgsql
+\c :db :u
+drop schema if exists s cascade;
+create schema s;
+create table s.t(k serial primary key, v int not null);
+insert into s.t(v) select generate_series(0, 49, 5);
+
+create function s.f()
+  returns table(z text)
+  set search_path = pg_catalog, pg_temp
+  language plpgsql
+as $body$
+declare
+  cur cursor(k_lo int, k_hi int) for (
+    select k, v
+    from s.t
+    where k between k_lo and k_hi
+    order by k);
+begin
+  assert pg_typeof(cur)::text = 'refcursor';
+  for rec in cur(2, 4) loop
+    assert pg_typeof(rec)::text = 'record';
+    z := rpad(rec.k::text, 10)||rec.v::text;      return next;
+  end loop;
+  z := '';                                        return next;
+  for rec in cur(7, 9) loop
+    z := rpad(rec.k::text, 10)||rec.v::text;      return next;
+  end loop;
+end;
+$body$;
+
+select s.f();
+```
+
+This is the result:
+
+```output
+ 2         5
+ 3         10
+ 4         15
+ 
+ 7         30
+ 8         35
+ 9         40
+```
+
+Notice these points:
+
+- With this form of the _query for loop_ the _loop variable_ is an implicitly declared record. (The first _assert_) confirms this.) You cannot use, say, two ordinarily declared _int_ variables like _k_out_ and _v_out_, even if this would serve your purposes better.
```

Comment:
Changed it to this:

> With this form of the _query for loop_, the _loop variable_ is an implicitly declared record. (The first _assert_ confirms this.) You cannot use, say, two ordinarily declared _int_ variables, like _k_out_ and _v_out_, even if this would serve your purposes better.

**Fixed.**